### DISCUSSION TITLE
How to use Morphology?

### DISCUSSION THREAD

**Discussion by imperator-maximus** [dlemstra/magick-wasm#107] [Q&A] [answered]
I just cannot import MorphologyMethod from "@imagemagick/magick-wasm";
and so I do not know how to continue. I also search the whole source but I cannot find any example. 
Thank you!

**Comment by dlemstra** [comment, marked as answer]
This not yet available in the current API but I just pushed a patch to make this available in the next release.

**Reply by imperator-maximus** [reply]
Thank you very much for quick reply and solution! I will try to build current version myself.

**Comment by imperator-maximus** [comment]
maybe you could help a bit to build it - would be awesome. 
I have:
```
npm run build

> @imagemagick/magick-wasm@0.0.23 build
> vite build && dts-bundle-generator src/index.ts -o dist/index.d.ts --external-inlines @dlemstra/magick-native

vite v4.4.7 building for production...
✓ 99 modules transformed.
dist/index.mjs  351.17 kB │ gzip: 69.33 kB
dist/index.umd.js  254.64 kB │ gzip: 62.61 kB
✓ built in 739ms
Compiling input files...
Processing src/index.ts
Writing src/index.ts -> dist/index.d.ts
Checking generated files...
Done in 4.17s
```
this looks great. 
with "npm prepare" the magick.wasm will be copied in that dist folder as well. also good.

But if I want to use it with same old code as before (not using new Morphology  - just some scaling) I am getting now 
![image](https://github.com/dlemstra/magick-wasm/assets/4406586/a9b603bd-1c30-4abb-8e2f-0764dd6241bc)
Any idea what I am doing wrong here? thank you!

**Reply by dlemstra** [reply]
This is probably something in your local setup? For example how you are referencing this. Not really sure how I can help you with this.

**Comment by imperator-maximus** [comment]
yes you are right I moved for new version some things and forgot to change path in initializeImageMagick() call. Works now.
Thank you!

**Comment by imperator-maximus** [comment]
I made some tests now and Morphology is working but it is up to 30 times slower than on same computer on command line:
```
time convert maske.png  -morphology Dilate:20 Square out.png
real    0m0.996s

```
compared to:
```
var settings = new MorphologySettings(MorphologyMethod.Dilate,Kernel.Square)
settings.iterations=20
return await ImageMagick.readFromCanvas(canvas, async image => {
    console.time("morphology")
    image.morphology(settings)
    console.timeEnd("morphology")
```
 this is 34 sec in browser (same computer). 
 I expect it slower but 34 times does not seem to be right here?
It also scales weird: with one iteration the difference is 5 times.

**Reply by dlemstra** [reply]
What happens when you do `new MorphologySettings(MorphologyMethod.Dilate,Kernel.Square, '20')` instead of setting the iterations?

**Comment by imperator-maximus** [comment]
yes tried this as well but that was much slower (had to close tab after some minutes). test image was 4500x2500 pixels in size 
![maske](https://github.com/dlemstra/magick-wasm/assets/4406586/b1078710-4943-4162-bd5d-cda863303b63)
btw. 
In the meantime I found another solution for my special use case by scaling down the image, doing that operation and scaling up again.

**Comment by dlemstra** [comment]
One reason for a longer duration could be that this all runs on a single thread and your command line uses multi threading.
